contracts/roles.js

```diff
@@ -0,0 +1,917 @@
+    params.processQueryLimit = api.BigNumber(processQueryLimit).toNumber();
+  }
+  await api.db.update('params', params);
+};
+
+async function updateCandidateWeight(id, deltaApprovalWeight, deltaToken = null) {
+  const candidate = await api.db.findOne('candidates', { _id: id });
+  if (candidate) {
+    if (deltaToken) {
+      const role = await api.db.findOne('roles', { _id: candidate.roleId });
+      const inst = await api.db.findOne('instances', { _id: role.instanceId });
+      if (inst.voteToken !== deltaToken.symbol) return true;
+    }
+    candidate.approvalWeight = {
+      $numberDecimal: api.BigNumber(candidate.approvalWeight.$numberDecimal)
+        .plus(deltaApprovalWeight),
+    };
+    await api.db.update('candidates', candidate);
+
+    const role = await api.db.findOne('roles', { _id: candidate.roleId });
+    role.totalApprovalWeight = {
+      $numberDecimal: api.BigNumber(role.totalApprovalWeight.$numberDecimal)
+        .plus(deltaApprovalWeight),
+    };
+    await api.db.update('roles', role);
+
+    return true;
+  }
+  return false;
+}
+
+actions.createInstance = async (payload) => {
+  const {
+    voteToken, candidateFee, isSignedWithActiveKey,
+  } = payload;
+
+  const params = await api.db.findOne('params', {});
+  const { instanceCreationFee } = params;
+
+  // eslint-disable-next-line no-template-curly-in-string
+  const utilityTokenBalance = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'" });
+
+  const authorizedCreation = api.BigNumber(instanceCreationFee).lte(0) || api.sender === api.owner
+    ? true
+    : utilityTokenBalance && api.BigNumber(utilityTokenBalance.balance).gte(instanceCreationFee);
+
+  if (api.assert(authorizedCreation, 'you must have enough tokens to cover the creation fee')
+    && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
+    if (candidateFee) {
+      if (!api.assert(typeof candidateFee === 'object'
+        && typeof candidateFee.method === 'string'
+        && FeeMethod.indexOf(candidateFee.method) !== -1
+        && typeof candidateFee.symbol === 'string'
+        && typeof candidateFee.amount === 'string' && api.BigNumber(candidateFee.amount).gte(0), 'invalid candidateFee properties')) return;
+      const feeTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: candidateFee.symbol });
+      if (!api.assert(feeTokenObj && api.BigNumber(candidateFee.amount).dp() <= feeTokenObj.precision, 'invalid candidateFee token or precision')) return;
+    }
+
+    const voteTokenObj = await api.db.findOneInTable('tokens', 'tokens', { symbol: voteToken });
+    if (!api.assert(voteTokenObj && voteTokenObj.stakingEnabled, 'voteToken must have staking enabled')) return;
```

Comment:
I don't think we should ever allow there to be a potentially unbounded number of operations happening when there's a staking update, there should always be some well-defined upper limit that we are comfortable with.  Otherwise there should be a mechanism to split update processing across multiple blocks, with intermediate state being stored between blocks.